### ISSUE TITLE
Lots  of `b-#{bid}-notify` keys with no expiration in my Redis instance 

### ISSUE THREAD

**Issue by delphaber** [sidekiq/sidekiq#6390]
Rails version: 6.1
Sidekiq / Pro / Enterprise version(s): BASE 6.5.12 / PRO 5.5.8

Surely I'm doing something wrong, but can someone please point me to the right direction? My redis instance is full of keys like `b-AG-P1Ko4ZMeRsQ-notify` with `1` as value and `-1` as TTL (aka, no expiration set).

I skimmed the sidekiq-pro source code and it seems like that `b-#{bid}-notify` keys should have an expiration, so I really don't know why I don't have any expiration at all.

Is it because I use `retry: false` for Sidekiq jobs inside batches?

The "Batch" page in the admin interface shows no batches pending or currently in progress.

**Comment by mperham** [comment]
I'm not aware of any bugs/leaks of this particular batch data element. Have you recently upgraded from an older Pro version? How long has this Redis dataset been in use? Perhaps this data is from long ago?

What you can do is update the expiry for those elements to expire in 30 days. Here's an generic example of a script to do so:

https://github.com/sidekiq/sidekiq/blob/main/Changes.md#412

You don't want to expire everything immediately because we aren't sure which batches are still alive and which are long gone so change that 60 to 2592000 (=30*24*60*60). If you still have a lot of these elements in 30 days or if you see new notify elements with a -1 TTL, open a new issue and we'll debug further.

**Comment by delphaber** [comment]
I've upgraded sidekiq-pro from 5.3.1 to 5.5.8 in October 2022.

> How long has this Redis dataset been in use? Perhaps this data is from long ago?

We started using sidekiq-pro in February 2021 (v. 5.2.1). I have a gut feeling that the number of `b-#{bid}-notify` with no expiration date is increased recently.

> What you can do is update the expiry for those elements to expire in 30 days. Here's an generic example of a script to do so:
> https://github.com/sidekiq/sidekiq/blob/main/Changes.md#412

Thanks! I'll try this :) I'll keep you updated in case new keys with no expiration appear in Redis

**Comment by delphaber** [comment]
I figured it out, I suppose :)

In the `:complete` callback for a sidekiq-pro batch I do `batch_status.delete`. That's because I don't really care about failed batches. It's totally OK to do not successfully finish a batch, that's why I also use `retry: false` on batches jobs.

So, doing `batch_status.delete` in the `:complete` callbacks leaves a `b-#{bid}-notify` key in redis with TTL: -1 (no expiration).

Today I've learned about the `:death` callback and I tried to move the `batch_status.delete` from the `:complete` to the `:death` callback and it removed all of the batch related keys from redis!

My question for you @mperham is: is it OK to use the `:death` callback for this cleanup operation? Otherwise, my only option left is to run a daily cronjob and cycle over `Sidekiq::Batch::DeadSet.new.each` and delete those batch statuses.
